sway-core/src/asm_generation/fuel/functions.rs

```diff
@@ -205,15 +207,46 @@ impl<'ir> FuelAsmBuilder<'ir> {
         // Compile instructions.
         let mut warnings = Vec::new();
         let mut errors = Vec::new();
-        for block in function.block_iter(self.context) {
-            self.insert_block_label(block);
-            for instr_val in block.instruction_iter(self.context) {
-                check!(
-                    self.compile_instruction(&instr_val, func_is_entry),
-                    return err(warnings, errors),
-                    warnings,
-                    errors
-                );
+
+        if let ProgramKind::Predicate = self.program_kind {
+            // For predicates, simply respect the original order in which the blocks are generated,
+            // assuming no CFG-altering passes have been run (e.g. simplify-cfg). This is temporary
+            // until restrictions on predicates are relaxed (namely allowing them to jump
+            // backwards).
+            for block in function.block_iter(self.context) {
+                self.insert_block_label(block);
+                for instr_val in block.instruction_iter(self.context) {
+                    check!(
+                        self.compile_instruction(&instr_val, func_is_entry),
+                        return err(warnings, errors),
+                        warnings,
+                        errors
+                    );
+                }
+            }
+        } else {
+            // For non-predicates, traverse the control flow graph using BFS. This is important to
+            // make sure that all values are inserted into `self.reg_map` in the right order,
+            // especially after CFG-altering optimization passes have been run.
+            let mut block_queue = VecDeque::new();
+            block_queue.push_back(function.get_entry_block(self.context));
+            let mut visited: HashSet<String> = HashSet::new();
+            while !block_queue.is_empty() {
+                let block = block_queue.pop_front().unwrap();
+                if visited.insert(block.get_label(self.context)) {
```

Comment:
```suggestion
            let mut visited: HashSet<Block> = HashSet::new();
            while !block_queue.is_empty() {
                let block = block_queue.pop_front().unwrap();
                if visited.insert(block) {
```
Did you try this?  A bit cheaper and simpler if it works.

Reply:
Super simple change otherwise, nice!